### PR TITLE
feature(pp): make pipeline parallelism first-class on the vLLM-native decode path

### PR DESCRIPTION
**Pull request by rebel-ykchoi**
### 🚀 Summary of Changes

> *What does this PR do? What feature, fix, or improvement does it bring?*

#### Summary

Pipeline parallelism (`pipeline_parallel_size > 1`) on the vLLM-native path
(`VLLM_RBLN_USE_VLLM_MODEL=1`) worked for plain decode but broke down the moment
decode batching, speculative decoding, or disaggregated serving entered the
picture: a step's decode batch could overflow the compiled per-stage bucket,
spec-decode misclassified the prefill/decode phase inconsistently across ranks,
and the KV connector saved only part of the model's layers. This PR reorganizes
the scheduler/runner so PP is a first-class, correct, and efficient mode across
all three.

The work is one story told in four steps: give the scheduler a single PP-aware
decode-admission budget, make speculative decoding rank-consistent on top of it,
consolidate the prefill/decode phase authority onto the scheduler (removing the
now-dead per-rank classifier), and turn the batch-balancing behavior on by
default for PP.

##### 1. PP-aware decode-batch admission (scheduler)
Under PP the engine keeps one decode microbatch in flight per stage, so a single
step's decode batch must fit the compiled per-stage bucket
(`max_num_seqs // pp_size`). Decode requests are admitted from two places — the
running queue and, in disaggregated serving, requests whose KV transfer just
completed — so admission is centralized into **one per-step budget** shared by
both loops (`DecodeAdmissionController` / `DecodeBatchBudget` in
`v1/core/utils.py`):
- Hard cap = the compiled bucket size on every decode that joins the batch
  (exceeding it leaves the runner without a matching bucket → crash).
- Optional **demand spreading** across stages (`ceil(demand / pp_size)`) so no
  stage idles after a pipeline drain.
- Requests that become decode-ready outside the demand snapshot (full local
  prefix-cache hit, resume-after-eviction) are gated by the hard cap only.
- The admitted count stays consistent when the batch shrinks (cleared on
  whole-batch eviction, decremented on preemption of an admitted decode).

`pp_size == 1` degenerates to `max_num_seqs` — unchanged.

##### 2. Speculative decoding under PP (scheduler + runner)
Enables spec decode (ngram, EAGLE, MTP, …) together with `pp_size > 1`. RBLN runs
synchronous scheduling with `pp_size` microbatches in flight, so a verify step —
whose query spans `num_spec_tokens + 1` positions and whose anchor/accepted
tokens are produced only on the last PP rank — must stay bit-consistent on every
non-last rank one step later. Three mechanisms:
- **Uniform per-query-length bookkeeping** — the prefill/decode phase is stamped
  once on `RBLNSchedulerOutput` by the scheduler and read by every rank (never
  re-derived per-rank), PP inter-stage tensors are sized per `(bucket, query_len)`,
  and non-last ranks write propagated tokens from `num_tokens_no_spec`.
- **Anchor-reconciled scheduling** — `should_defer_spec_step` admits a verify only
  when its base token is reconciled, deciding on the base count (not the
  draft-inflated `num_new_tokens`). Gated on spec being enabled → non-spec decode
  is untouched.
- **Multi-accept token propagation** — the scheduler extends a lagging rank's
  `new_token_ids` payload back by `num_spec_tokens`; the runner writes by absolute
  position (idempotent over mis-speculated slots).

**KV-connector finalization is made PP-correct**: the deferred per-step finalize
(so draft-token KV is saved) runs in the sample path that only the last rank
reaches, so it is gated on the last rank — non-last ranks finalize in-context, so
every rank still saves its own layers' KV.

##### 3. Scheduler as the single phase authority (runner cleanup)
With `RBLNSchedulerOutput` now the unified type every step carries, the legacy
per-rank `is_prefills()` classifier (which diverged across PP ranks under spec
decode) is removed along with its test and a dangling comment. The defensive
`isinstance`/`getattr` fallbacks that hedged for a plain `SchedulerOutput` are
dropped: RBLN-only helpers are narrowed to `RBLNSchedulerOutput`, and the
base-override methods sit behind a single boundary assert that both guards the
invariant and narrows the type for direct field access.

##### 4. PP decode-batch balancing on by default
`VLLM_RBLN_PP_BALANCE_DECODE_BATCH` now defaults to **on**. The dynamic cap only
ever caps *lower* than the static one (never overflows the bucket), is floored at
1, matches the static cap at saturation, and is a **no-op for `pp_size == 1`** —
so non-PP configs are unaffected. Remains as an escape hatch: set `=0` to force
the static cap.

#### Testing

- **Unit** — new/extended coverage for the backfill/slide query-window math
  (`test_query_backfill.py`), the anchor-reconciled deferral predicate and
  multi-accept payload (`test_rbln_model_runner.py`), the PP decode-admission
  budget and demand-spread contrast (`test_scheduler.py`), and PP disaggregation
  (`test_pd_disaggregation.py`).
- `pre-commit run --hook-stage manual` clean (ruff, mypy 3.10–3.13, typos,
  license); full unit suites for runner/scheduler/pd green.

#### Risk / compatibility

- `pp_size == 1` and non-spec decode paths are untouched (predicate gated on spec;
  cap degenerates to `max_num_seqs`; balancing is a no-op).
- The default flip is confined to PP; the escape hatch preserves the old behavior.

#### Affected modules

Platform, Scheduler, Runner.

---

### 📌 Related Issues / Tickets

* Resolves #
* Related to #

---

### ✅ Type of Change

* [ ] 🚀 Release (`release`)
* [x] ✨ Feature (`feature`)
* [ ] 🧠 Model support (`model`)
* [ ] 🧬 Core engine changes (`core`)
* [ ] 🛠 Bug fix (`fix`)
* [ ] ⚙️ Performance improvement (`perf`)
* [ ] 🔁 Refactor or code cleanup (`refactor`)
* [ ] 📄 Documentation (`docs`)
* [ ] ❓ Other (`other`): please describe

---

### 🧪 How to Test

1. Run `...`
2. Verify output: `...`
3. Edge case tested: `...`

---

### 📸 Screenshots / Logs (if applicable)

---

### 📋 Checklist

* [ ] PR title follows Conventional Commits format
* [ ] This PR is linked to an existing issue
* [ ] The test method is described, and the expected result is clearly stated
* [ ] Relevant documentation has been updated (if applicable)

---

### 💬 Notes

---

### CODE REVIEW

File: vllm_rbln/v1/core/rbln_scheduler.py [hunk 1]
```diff
@@ -288,6 +321,21 @@ def _decide_spec_slide(self, request: Request, new_n: int) -> tuple[int, bool]:
...
+        = running decodes + remote-KV requests ready to be admitted (transfer
+        complete, awaiting promotion). Including the ready remote-KV gives the
+        dynamic cap headroom to ramp the decode batch on the P/D-disaggregated
+        decode side; running-only would stall the ramp. Demand is invariant
+        under admission (promoting a ready remote-KV moves it from ready ->
+        running), so this snapshot is exact even as the running/ready split
+        shifts during the waiting loop. Evaluated only when PP balancing is on.
+        """
+        num_running_decodes = sum(1 for r in self.running if not is_prefill(r))
+        num_ready_remote_kv = len(self.finished_recving_kv_req_ids)
```
Comment (rebel-jinhwan) [inline vllm_rbln/v1/core/rbln_scheduler.py:336]: LMCache MP Connector return `load_kv_async=True` even for partial prefix hit. Those requests are still prefill request.

Comment (rebel-ykchoi) [reply vllm_rbln/v1/core/rbln_scheduler.py:336]: This is intentional. `_decode_demand` is a deliberately *inclusive* count of
requests that may become decodes (running decodes + ready remote-KV), used only
to size the PP-balancing **soft/spreading cap** `ceil(demand / pp_size)`. It does
not — and need not — resolve prefill-vs-decode at count time; that determination
is deferred to the admission loop below, per request:

- The soft cap gates only remote-KV promotions (`apply_soft_cap =
  request.status == WAITING_FOR_REMOTE_KVS`). A gated request stays in
  `WAITING_FOR_REMOTE_KVS` and is re-evaluated next step — it is not flipped and
  mis-routed through local prefill (see the comment at the admission site).
- So if such a finished-recving request is in fact a prefill (e.g. the LMCache
  MP partial-hit `load_kv_async=True` case you point out), it is handled as a
  prefill downstream. RBLN has no mixed batching, so a prefill step evicts the
  decode batch and the per-step decode budget is reset — the transient
  over-inclusion in the demand snapshot doesn't leak into the decode accounting.
- The **hard cap** (the compiled decode bucket, `max_num_seqs // pp_size`),
  enforced unconditionally by `can_admit()`, is the correctness limit and is
  independent of `_decode_demand`. An over-count can only make the soft cap
  slightly more permissive (weaker spreading for that step) — never breach the
  bucket. And it self-corrects: once the request enters `running`, the next
  step's snapshot reflects reality.

So the partial-hit-prefill case is covered by design — the demand snapshot is an
approximate spreading input, and the actual routing/accounting is exact at
admission time.
